mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -433,6 +433,92 @@ void SliceOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<ConcatSliceOptimization>(context);
 }
 
+struct MinToClampOptimization : public OpRewritePattern<tosa::MinimumOp> {
+  using OpRewritePattern<tosa::MinimumOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::MinimumOp op,
+                                PatternRewriter &rewriter) const override {
+
+    DenseElementsAttr constant;
+    if (!matchPattern(op.getInput2(), m_Constant(&constant)) ||
```

Comment:
What if input1 is the splat constant and input2 is the generic input? Can we also rewrite that case?

Reply:
In my mind, MLIR had a canonicalization that would always use the second operand as constant. Let me check.

Reply:
Confirmed MLIR has a canonicalization to move the constant to the second operand.
```
func.func @clamp_minimum_f32(%arg0: tensor<4xf32>) -> tensor<4xf32> {
    // CHECK: "tosa.clamp"(%arg0) <{max_fp = 6.000000e+00 : f32, max_int = 6 : i64, min_fp = -3.40282347E+38 : f32, min_int = -2147483648 : i64}
  %0 = "tosa.const"() <{value = dense<6.0> : tensor<1xf32>}> : () -> tensor<1xf32>
  %1 = "tosa.minimum"(%arg0, %0) : (tensor<4xf32>, tensor<1xf32>) -> tensor<4xf32>
  return %1 : tensor<4xf32>
}

func.func @clamp_minimum2_f32(%arg0: tensor<4xf32>) -> tensor<4xf32> {
    // CHECK: "tosa.clamp"(%arg0) <{max_fp = 6.000000e+00 : f32, max_int = 6 : i64, min_fp = -3.40282347E+38 : f32, min_int = -2147483648 : i64}
  %0 = "tosa.const"() <{value = dense<6.0> : tensor<1xf32>}> : () -> tensor<1xf32>
  %1 = "tosa.minimum"(%0, %arg0) : (tensor<1xf32>, tensor<4xf32>) -> tensor<4xf32>
  return %1 : tensor<4xf32>
}
```
both get canonicalized:
```
func.func @clamp_minimum_f32(%arg0: tensor<4xf32>) -> tensor<4xf32> {
  %0 = "tosa.clamp"(%arg0) <{max_fp = 6.000000e+00 : f32, max_int = 6 : i64, min_fp = -3.40282347E+38 : f32, min_int = -2147483648 : i64}> : (tensor<4xf32>) -> tensor<4xf32>
  return %0 : tensor<4xf32>
}
func.func @clamp_minimum2_f32(%arg0: tensor<4xf32>) -> tensor<4xf32> {
  %0 = "tosa.clamp"(%arg0) <{max_fp = 6.000000e+00 : f32, max_int = 6 : i64, min_fp = -3.40282347E+38 : f32, min_int = -2147483648 : i64}> : (tensor<4xf32>) -> tensor<4xf32>
  return %0 : tensor<4xf32>
}
```